pdns/zoneparser-tng.cc

```diff
@@ -272,6 +272,38 @@ pair<string,int> ZoneParserTNG::getLineNumAndFile()
     return {d_filestates.top().d_filename, d_filestates.top().d_lineno};
 }
 
+void fixTxtRecord(string &txt) {
+  bool in_quote = false;
+  bool needs_end_quote = false;
+
+  // Bind automatically puts quotes around text blocks that don't have them
+  for( string::size_type i = 0; i < txt.size(); ++i ) {
+    if( in_quote ) {
+      if( needs_end_quote && txt[i] == ' ' ) {
+        txt.insert(i, "\"");
+        ++i;
+        in_quote = false;
+      }
+      else if( !needs_end_quote && txt[i] == '"' ) {
+        in_quote = false;
+      }
+      else if( txt[i] == '\\' ) {     // escape - skip the next character
+        ++i;
+      }
+    } else if( txt[i] != ' ' ) {      // start of a new string
+      in_quote = true;
+      if( txt[i] != '"' ) {
+        txt.insert(i, "\"");
+        ++i;
+        needs_end_quote = true;
+      } else
+        needs_end_quote = false;
+    }
+  }
+  if( in_quote && needs_end_quote )
+    txt.append("\"");
```

Comment:
Wouldn't it be better to reject bad input than to try correcting it and likely screw things up?

Reply:
I agree it would be best to reject bad input, however not quoting your TXT is not bad input as it works fine in named.

Reply:
named is not the definition of master file syntax; [RFC 1035](https://tools.ietf.org/html/rfc1035#section-5.1) is. If the former allows unterminated \<character-string>s like e.g. `TXT "foo`, then it's going off-script.

Reply:
It's not unterminated as much as implicitly quoting unquoted sections eg `TXT foo`

Reply:
Which, looking at the RFC is how it should be so pdns is currently in regression to this...

```
<character-string> is expressed in one or two ways: as a contiguous set
of characters without interior spaces, or as a string beginning with a "
and ending with a ".  Inside a " delimited string any character can
occur, except for a " itself, which must be quoted using \ (back slash)
```